### ISSUE TITLE
StatefulSets for Prometheus/Alertmanager (Single Instance Only)

### ISSUE THREAD

**Issue by gardener-robot-ci-1** [gardener/gardener#53]
By @afritzler: I just wanted to open the discussion here why we are using stateful sets for the alertmanager and prometheus deployments. IMO we should switch to single pod replicasets. That is what you see out there in the wild when looking for example manifests for prometheus.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [rfranzke](https://github.com/rfranzke)**
_Wednesday Nov 29, 2017 at 11:27 GMT_

----

@i068969 please shed some light here

**Comment by gardener-robot-ci-1** [comment]
**Comment by [mvladev](https://github.com/mvladev)**
_Wednesday Nov 29, 2017 at 12:28 GMT_

----

`ReplicaSet` should only be used by other controllers. You probably meant `Deployment`? I didn't want to add manual persistent volume management (deletion to be specific). So `StatefulSet` felt as the easiest solution.

AlertManager requires for every replica to talk to other replicas. I believe only `StatefulSet` [supports](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id) that.

We can switch to `Deployments` and `PVC`, but we have to generate owner reference pointing to the shoot's namespace for easy deletion

**Comment by gardener-robot-ci-1** [comment]
**Comment by [rfranzke](https://github.com/rfranzke)**
_Monday Dec 04, 2017 at 06:54 GMT_

----

Hey @i068969, I didn't get the point regarding the PV management/the argument regarding the deletion. When deleting a Shoot cluster, we delete the whole namespace within the Seed cluster. That will recursively delete all objects (and also the AlertManager/Prometheus PVCs).

And what exactly was the argument for two replicas while we have only one for all other deployments (especially the control plane)?

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Wednesday Dec 27, 2017 at 09:39 GMT_

----

Two remarks:
* If a StatefulSet is deleted, it doesn't delete the PVCs, so StatefulSets don't relieve us from taking care of the PVCs/PVs
* However, as pointed out, we delete the namespace and now it's good that PVs are not bound to the namespace (this time at least), because if K8s/we would accidentally delete the PV before the PVC (which is deleted when the namespace is deleted), we would end up with orphan volumes as backing volumes are not deleted if the PV is deleted (which is nuts)

In other words:
* StatefulSets are not helpful when it comes to deletion
* The PV handling is crap in Kubernetes and we must make sure to work with PVCs at all times and to avoid any PV pitfalls in cleanup

Question remains, whether we switch to deployment + PVC or not? Remove the discussion label or close this ticket for good?

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Wednesday Dec 27, 2017 at 09:43 GMT_

----

P.S.: I can't find the ticket, but wasn't there a ticket/discussion around the two alert manager instances? If not and we only talked verbally about the subject matter in the context of this ticket, I suggest to switch to one replica like for everything else or motivate why we run two alert managers?

We need to understand whether the single-pod philosophy is acceptable or not. If not, I see greater harm in running etcd or the other main control plane components as single pods than in the alert manager. Or, the other way around, if it's ok to have them as single pods, why would we need two alert managers?

**Comment by gardener-robot-ci-1** [comment]
**Comment by [I820491](https://github.com/I820491)**
_Tuesday Jan 02, 2018 at 12:27 GMT_

----

I think the alertmanager discussion was verbally only. IMO we should switch back to deployment + pvc and have the replica set to 1. I haven't found anybody on github who is deploying Prometheus/Alertmanager as StatefulSets. A good baseline are the helm charts for Prom [1].

[1] https://github.com/kubernetes/charts/tree/master/stable/prometheus/templates

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vlerenc](https://github.com/vlerenc)**
_Wednesday Jan 03, 2018 at 07:28 GMT_

----

Well, not seeing other people doing it this way is an indication, but no strong reasoning. Other people weren't deploying the control plane in the way we do it, but here we are with Gardener.

Anyway, I also don't see arguments by Martin (the one given isn't holding true - deletion of PVCs isn't happening automatically).

In the end, I remain open to both views. That means, who touches it (the code) decides. My personal preference is slightly towards deployments + PVCs as that's what we are doing so far and we have seen some "special" behaviour in StatefulSets. In that sense, different code paths are used here and my feeling is, we lack sufficient experience with them.

What I very much like to change is the 2 x AlertManager deployment. If we need it, we must rethink the control plane approach we took so far. If not, it's a waste of resources and not logical to keep 2 instances running.

**Comment by gardener-robot-ci-1** [comment]
**Comment by [vasu1124](https://github.com/vasu1124)**
_Wednesday Jan 03, 2018 at 11:21 GMT_

----

a) StatefulSets are required if the internal workload need a sticky identity,
b) StatefulSets have an additional feature, where volumeClaimTemplates can be used to dynamically create PVCs per requested number of replicas (and bind them to ordinal ordered pods),
c) Deployments/ReplicaSets treat volumes as stateless.

Prometheus and Alertmanager to my knowledge do not need sticky identities, but nevertheless, if the number of replicas needs to be increased to >1 for HA or redundancy (and the volume is of type read-write-one), then afaik Deployment/ReplicaSets are not an option.

**Comment by vlerenc** [comment]
@afritzler, @mvladev, @vasu1124, @rfranzke What's the status of this ticket?

I suggest to keep the stateful sets (we have them anyways for etcd) and don't change that. However, I suggest to scale down the alert manager to one instance as we run everything with one instance.

**Comment by rfranzke** [comment]
I am of the same opinion, let's stay with StatefulSets.
Reduction to 1 replica for AlertManager has been done with 5f2815968be90b81f9768f1fdcb51be4fa133cfe.

**Comment by vlerenc** [comment]
Thanks!
